lib/ask/runtime/survey.ex

```diff
@@ -43,138 +43,103 @@ defmodule Ask.Runtime.Survey do
+    end
+    {:end, updated_respondent}
   end
 
-  def update_respondent(%Respondent{} = respondent, {:rejected, session, timeout}) do
-    timeout_at = Respondent.next_actual_timeout(respondent, timeout, SystemTime.time.now)
-    respondent
-      |> Respondent.changeset(%{state: "rejected", session: Session.dump(session), timeout_at: timeout_at})
-      |> Repo.update!
+  def handle_session_step({:failed, respondent}, _, offline) do
+    updated_respondent = failed_session(respondent, offline)
+    {:end, updated_respondent}
   end
 
-  def update_respondent(%Respondent{} = respondent, :failed) do
+  def failed_session(respondent, offline) do
     session = respondent.session |> Session.load
     mode = session.current_mode |> SessionMode.mode
     old_disposition = respondent.disposition
     new_disposition = Flow.failed_disposition_from(respondent.disposition)
 
-    Session.log_disposition_changed(respondent, session.current_mode.channel, mode, old_disposition, new_disposition)
-
-    respondent
-    |> Respondent.changeset(%{state: "failed", session: nil, timeout_at: nil, disposition: new_disposition})
-    |> Repo.update!
-    |> RespondentDispositionHistory.create(old_disposition, mode)
-  end
-
-  def update_respondent(%Respondent{} = respondent, :stopped, disposition, _) do
-    session = respondent.session |> Session.load
-    update_respondent_and_set_disposition(respondent, session, nil, %{disposition: disposition, state: "failed", session: nil, timeout_at: nil, user_stopped: true})
-  end
-
-  def update_respondent(%Respondent{} = respondent, {:ok, session, timeout}, nil, now) do
-    effective_modes = respondent.effective_modes || []
-    effective_modes =
-      if session do
-        mode = Ask.Runtime.SessionMode.mode(session.current_mode)
-        Enum.uniq(effective_modes ++ [mode])
-      else
-        effective_modes
-      end
-
-    timeout_at = Respondent.next_actual_timeout(respondent, timeout, now)
-    respondent
-    |> Respondent.changeset(%{state: "active", session: Session.dump(session), timeout_at: timeout_at, language: session.flow.language, effective_modes: effective_modes})
-    |> Repo.update!
-  end
-
-  def update_respondent(%Respondent{} = respondent, {:ok, session, timeout}, disposition, _) do
-    timeout_at = Respondent.next_actual_timeout(respondent, timeout, SystemTime.time.now)
-    update_respondent_and_set_disposition(respondent, session, timeout, %{session: Session.dump(session), timeout_at: timeout_at, disposition: disposition, state: "active"})
+    updated_respondent = respondent_updates(:failed, respondent, offline)
+    Session.log_disposition_changed(updated_respondent, session.current_mode.channel, mode, old_disposition, new_disposition)
+    disposition_changed(updated_respondent, session, old_disposition, offline) #TODO: this knows that the line before changed the disposition
   end
 
-  def update_respondent(%Respondent{} = respondent, :end, reply_disposition, _) do
```

Comment:
Is this case being considered? Was it deprecated somehow? The `respondent_updates(:end, #...` definition doesn't look anything like this one.

Reply:
Yes, this was one of the functions that was doing way-much-more than just updating the respondent. If you take a look at 
```elixir 
handle_session_step({:end, reply, respondent}...
``` 
you will find out that the logic that is left over `respondent_updates` is there